### PR TITLE
makes java_image rule main_class parameter optional

### PR DESCRIPTION
**Pull request by michaelschiff**
Allows additional jvm_flags to be provided at runtime. Example of this in kubernetes would look something like:

```
      - name: myapp
        image: myapp
        args:
          - '-Xmx10g'
          - '-Xms10g'
          - 'com.mycompany.MyApp'
          - 'applicationArgument1'
          - 'applicationAgument2'
```
Based on some initial discussion here: https://github.com/bazelbuild/rules_docker/issues/548
The intent here is to allow users to build a single image with the java_image rule, but to have the ability to add JVM parameters at runtime based on the context in which they are running the image (size of pod, production vs test environment). This is preferable to having to build an image per combination of possible flags they would require. If there is a better way to supply these flags at runtime I am happy to discuss alternatives.

The idea in the issue linked above was based on possible variable expansion and requires images have a shell, which the distroless images do not.  The solution in this PR aims to be simple, backwards compatible, and compatible with images that do not include a shell

### CODE REVIEW

**Comment by michaelschiff** [comment]
@nlopezgi certainly - I'll add those tests now

**Comment by michaelschiff** [comment]
@nlopezgi Hi sorry for the delay - Using the presence or absence of `main_class` was simple, but made some use cases not possible. 

The call to `native.java_binary` will work without `main_class` as a parameter, but only if the package contains the word "java" and will use the `name` parameter as the value of the main class in the binary script.  This meant that if you wanted to use a partial entrypoint, but still compile additional sources at the call to `java_image` that your code needed to reside in a package that contained the word "java".  If the `name` value was incorrect, this does not prevent the image from running properly, but the binary script that is generated cannot be executed (due to the incorrectly assumed value of the main class).  I felt like this made too many assumptions about the structure of projects and relied on naming conventions in an unpleasant way.  I thought about requiring that `main_class` be present if `srcs` were to be compiled, and only calling `native.java_binary` if there were sources to compile, but this meant that you could not compile additional sources and take advantage of a partial entrypoint.

Because of this, I chose to add an additional parameter `partial_entrypoint` that, when True, would yield an entrypoint that stops before the inclusion of the main class.  I wanted to avoid adding an additional parameter, but the default value of False means that the change is still backwards compatible.

I've also added a test for this change - It looks like all existing java_image tests are setup as image definitions in `testdata/BUILD` and are checked in `testing/e2e.sh`, so I've added my test here. I can add an additional test like the one you described if that is preferable.

**Comment by nlopezgi** [comment]
Hi @michaelschiff thanks for working on this PR. Unfortunately adding additional parameters to xx_image rules is a no go as we want to keep their API aligned with xx_binary rules. Lets step back and look at what you need and how to provide it in a way that makes more sense for these rules.

Let me ask a question to understand your use case better:
- If you had a similar need for a java application that did not run inside a container, how would you achieve this with a java_binary target?

(my speculative answer) I think the right way would be to provide this is via having several java_toolchain targets (https://docs.bazel.build/versions/master/be/java.html#java_toolchain) one for each set of flags you need, and then allowing users to pick the java_toolchain at runtime.

If this is the case, then the solution to this is to actually support proper use of the java_toolchain rule in the java_image rule (currently our entrypoint just hardcodes the jvm location, but all this info should come from the java_toolchain rule). The problem we might have atm is that java rules are not using the java toolchain correctly (its work in progress iiuc). We should be able to figure out a way to use the java_toolchain correctly here nevertheless. Let me know if this sounds like the right path to you as well and we talk more about details.

**Comment by michaelschiff** [comment]
Ah okay, I saw the `layers` parameter and figured some additional parameters may be tolerable, but I understand that API changes are a bigger deal.

> If you had a similar need for a java application that did not run inside a container, how would you achieve this with a java_binary target?

for this, `jvm_flags` actually works as expected.  The script that is generated by `java_binary` can be passed `--jvm_flags` as an argument at runtime and these flags are provided to the java runtime.

Toolchains does not look like the right answer, as this is for setting options for the java compiler.  I don't see a way to use java_toolchains to set parameters like the heap size for the actual binary you are generating, and would still be happening at java compile time, not runtime.

>  having several java_toolchain targets (https://docs.bazel.build/versions/master/be/java.html#java_toolchain) one for each set of flags you need, and then allowing users to pick the java_toolchain at runtime.
When you say "pick the java_toolchain at runtime" I take it you mean bazel runtime, i.e. when the image is being built.  The use case I am trying to solve for is at container runtime.  The end goal being to build a single container, and be able to pass in different jvm flags at container runtime.

**Comment by nlopezgi** [comment]
Thanks for the additional info.
It sounds like what we are missing is a script that uses the jvm_flags inside the container. Right now we create a very simple entrypoint for the java image:
https://github.com/bazelbuild/rules_docker/blob/master/java/image.bzl#L190
while bazel generates the complex script, which, as you mention, includes a way to resolve the jvm_flags properly (https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt#L35). I think we just might need to have a script that does the same thing as the one generated by java_binary, at least wrt jvm_flags, if we want to get parity on these features, and we would need to use that as the entrypoint, not /usr/bin/java. Do you think that might work?

**Comment by michaelschiff** [comment]
That would certainly work - but it would require that the image contain a shell, which the distroless images do not.  Though I suppose this "script" could be written in java to avoid the dependency on a shell.

```
# --debug               Launch the JVM in remote debugging mode listening
# --debug=<port>        to the specified port or the port set in the
#                       DEFAULT_JVM_DEBUG_PORT environment variable (e.g.
#                       'export DEFAULT_JVM_DEBUG_PORT=8000') or else the
#                       default port of 5005.  The JVM starts suspended
#                       unless the DEFAULT_JVM_DEBUG_SUSPEND environment
#                       variable is set to 'n'.
# --main_advice=<class> Run an alternate main class with the usual main
#                       program and arguments appended as arguments.
# --main_advice_classpath=<classpath>
#                       Prepend additional class path entries.
# --jvm_flag=<flag>     Pass <flag> to the "java" command itself.
#                       <flag> may contain spaces. Can be used multiple times.
# --jvm_flags=<flags>   Pass space-separated flags to the "java" command
#                       itself. Can be used multiple times.
# --singlejar           Start the program from the packed-up deployment
#                       jar rather than from the classpath.
# --print_javabin       Print the location of java executable binary and exit.
# --classpath_limit=<length>
#                       Specify the maximum classpath length. If the classpath
#                       is shorter, this script passes it to Java as a command
#                       line flag, otherwise it creates a classpath jar.
```
If we go with this approach we can aim for feature parity with the above set of flags

**Comment by nlopezgi** [comment]
That sounds like its the right approach. I'm slightly concerned there will be a high maintenance effort that will be required to keep a 'java version of the script' working well and on par with the Bazel shell version, but perhaps with the proper design this can be mitigated. How do you suggest we move forward (e.g., do you want to take a shot at building the functionality, want to hash out its design a bit more in a doc, other alterantives)

**Comment by michaelschiff** [comment]
@nlopezgi - So I've pushed another change as one last attempt at keeping things simple. I wanted to explore all of the simple options before we go down the path of writing a "java version" of the entrypoint script, because I agree with you that this exposes us to a potential maintenance headache.

This commit addresses the issue I mentioned earlier about the call to `native.java_binary` without a `main_class` parameter.  Instead of passing None, we will always provide a value for `main_class`, either the value provided by the user, or if it was none, the value of the binary_name.   This behavior is consistent with the resulting behavior of `native.java_binary` if the package contains "java" or "javatest", but without the package naming convention.  

The downside to this approach is that the script that is produced by the call to `native.java_binary` will use the binary_name as the name of the main class, which may not be correct.  However, this is consistent with the behavior or `native.java_binary` which will use the value of the `name` parameter if the `main_class` parameter is missing and the package name contains "java" or "javatest" (if the package name does not contain one of these strings it's an error).

I think this is acceptable for two reasons: First is the consistency between this behavior, and that of `java_binary`.  Second, the script output from `java_binary` is never actually called as part of the image (as we have already established), so we are not impacted by the potentially incorrect main_class name used in this script.

Doing this achieves my original goal of allowing users to pass additional jvm flags at runtime, in a backwards compatible way, and without any modification to the `java_image` API.

All that said, if this does not seem like the right approach to you I am happy to start on a design for the new entrypoint program we were discussing.

**Comment by nlopezgi** [comment]
kicked off presubmits. I'm concerned the solution makes the api between java_image and java_binary diverge, and that it does not solve the issue for other things the script does. I do think the right solution is to provide the equivalent of the script, but the fact that this seems hard to do (and even harder to maintain) poses significant questions as to the long term viability of maintaining the java_image sensibly on par with java_binary.

**Comment by michaelschiff** [comment]
> concerned the solution makes the api between java_image and java_binary diverge
The main divergence I can think of is that `java_binary` will allow you to omit the `main_class` param if your package includes java/javatest in the name, in which case it will guess the main_class based on the logic here (https://github.com/bazelbuild/bazel/blob/cd699abd858673658dbe7ee2133c9af934a3fda9/src/main/java/com/google/devtools/build/lib/rules/java/JavaCommon.java#L531).  `java_image` now allows you to omit the main_class regardless of package naming.  Keeping the resulting `main_class` value the same means getting as close to the logic of `determinePrimaryClass`, right now it could do better.  I still feel like this is in the realm of acceptability, since the binary produced by a `java_image` rule is unlikely to be called (but I can work to improve this).

>it does not solve the issue for other things the script does
though it does not provide the same flags, you should be able to produce most behaviors as you would if the script were the entrypoint.  In the case of `--debug` you are able to pass in the necessary flags (just like you can pass in additional jvm flags).  The same should be true for `--main_advice`.  `--singlejar` and `--print_javabin` are not really supported right now, and the handling of classpath length limits is a little different.  I think for most use cases this is sufficient. 

Long term, I agree that a more feature rich entrypoint is probably a good idea.  Seems like we should start that discussion outside of this PR.  What's the best place for this kind of design discussion?

**Comment by nlopezgi** [comment]
thanks for your clarifications. I'll opened https://github.com/bazelbuild/rules_docker/issues/574 about the larger issue of getting parity with the java stub script.

**Comment by michaelschiff** [comment]
👍 Ill address your comments and we can take the long-term discussion to #574

File: java/image.bzl [hunk 1]
```diff
@@ -264,10 +264,14 @@ def java_image(
     **kwargs: See java_binary.
   """
     binary_name = name + ".binary"
-
     native.java_binary(
         name = binary_name,
-        main_class = main_class,
+        # Calling java_binary with main_class = None will work if the package binary_name
+        # contains java or javatest.  In this case, the main_class is guessed to be the
```
[the remark this answers is not in the document]
Comment (michaelschiff) [reply java/image.bzl]: The error thrown by `java_binary` would look something like:
```
in java_binary rule <rule> cannot determine main class for launching (found neither a source file '<rule_name>.java', nor a main_class attribute, and package name doesn't include 'java' or 'javatests')
```
This is part of the definition of `java_rule` (more information about what it looks for here https://github.com/bazelbuild/bazel/blob/cd699abd858673658dbe7ee2133c9af934a3fda9/src/main/java/com/google/devtools/build/lib/bazel/rules/java/BazelJavaSemantics.java#L166 and https://github.com/bazelbuild/bazel/blob/cd699abd858673658dbe7ee2133c9af934a3fda9/src/main/java/com/google/devtools/build/lib/rules/java/JavaCommon.java#L531).

After the change in this PR, we will never pass `main_class = None` to `native.java_binary` (instead we pass the `binary_name` when a `main_class` value is not provided), so the user should never see this error.

File: java/image.bzl [hunk 2]
```diff
@@ -192,7 +192,7 @@ def _jar_app_layer_impl(ctx):
         "-cp",
         # Support optionally passing the classpath as a file.
         "@" + classpath_path if ctx.attr._classpath_as_file else classpath,
-    ] + jvm_flags + [ctx.attr.main_class] + args
+    ] + jvm_flags + ([ctx.attr.main_class] + args if ctx.attr.main_class != "" else [])
```
[the remark this answers is not in the document]
Comment (michaelschiff) [reply java/image.bzl]: In the construction of the entrypoint there is no guessing about `main_class`.  It is only added to the entrypoint if it was supplied by the user in the call to `java_image`.  If it was not supplied, the entrypoint construction will not include a main class (this is what allows you to specify additional jvm flags at runtime, followed by the main class (my first comment contains an example of what this looks like)).

The only place we are guessing the main class value is in our call to `native.java_binary`.  I can update the doc comment for the `java_image` rule to explain this behavior

**Comment by michaelschiff** [comment]
@nlopezgi yeah, no problem.  I'll add the extra test.

**Comment by michaelschiff** [comment]
@nlopezgi I added those tests - one for java image where the main_class was provided, and one where it was not.  Adding the test was easy, but it looks like the string comparison that used to check the entrypoint may not be working properly:

```
--- FAIL
Error: Image entrypoint [/usr/bin/java -cp /app/io_bazel_rules_docker/testdata/libjava_bin_as_lib.jar:/app/io_bazel_rules_docker/testdata/libjava_image_library.jar:/app/io_bazel_rules_docker/../com_google_guava_guava/jar/guava-18.0.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary examples.images.Binary] does not match expected entrypoint: [/usr/bin/java -cp /app/io_bazel_rules_docker/testdata/libjava_bin_as_lib.jar:/app/io_bazel_rules_docker/testdata/libjava_image_library.jar:/app/io_bazel_rules_docker/../com_google_guava_guava/jar/guava-18.0.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary examples.images.Binary]
ERROR: FAIL
```

File: tests/docker/configs/java_entrypoint.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+schemaVersion: 2.0.0
+
+metadataTest:
+  entrypoint: ["/usr/bin/java -cp /app/io_bazel_rules_docker/testdata/libjava_bin_as_lib.jar:/app/io_bazel_rules_docker/testdata/libjava_image_library.jar:/app/io_bazel_rules_docker/../com_google_guava_guava/jar/guava-18.0.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary.jar:/app/io_bazel_rules_docker/tests/docker/java_image.binary examples.images.Binary"]
```
Comment (nlopezgi) [inline tests/docker/configs/java_entrypoint.yaml]: you need to split lines at ' '
https://github.com/bazelbuild/rules_docker/pull/576

**Comment by nlopezgi** [comment]
almost there, tests should work is you split lines in the yaml file

File: tests/docker/configs/java_entrypoint.yaml (hunk 3 above)
Comment (michaelschiff) [reply tests/docker/configs/java_entrypoint.yaml]: got it - fixed

**Comment by ceason** [comment]
@michaelschiff did you try using the [`JAVA_TOOL_OPTIONS`](https://docs.oracle.com/javase/8/docs/platform/jvmti/jvmti.html#tooloptions) env var? Per your example this would look like:
```
      - name: myapp
        image: myapp
        args:
        - 'applicationArgument1'
        - 'applicationAgument2'
        env:
        - name: JAVA_TOOL_OPTIONS
          value: '-Xms10g -Xmx10g'
```

**Comment by michaelschiff** [comment]
@ceason hmm, I thought I had tried all of the avenues for setting these options through environment variables, but I must have missed this.  JAVA_TOOL_OPTIONS does work.

Since it can be disabled 
>Platforms may disable this feature in cases where security is a concern; for example, the Reference Implementation disables this feature on Unix systems when the effective user or group ID differs from the real ID

I think it makes sense to leave the ability to update the command entrypoint directly, but it certainly seems preferable to use `JAVA_TOOL_OPTIONS` when it is available.

@nlopezgi thoughts?

**Comment by ceason** [comment]
@michaelschiff at least in the context of linux, the only time your effective/real IDs can differ is via `setuid/setgid` (eg `sudo`). The `"jvm"_image` rules don't invoke java like that, so I think you should be safe there.

Your quoted reference made me curious though--I know k8s lets you specify a particular UID for a container with `securityContext.runAsUser`..but does k8s use `setuid` for that? I cooked up a pod to answer that question:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: busybox
spec:
  containers:
  - &c
    name: busybox
    image: busybox
    command: [/bin/sh, -c]
    args:
    - |
      while [ true ]; do
        echo "real=$(id -ur), effective=$(id -u)"
        sleep 2
      done
  - <<: *c
    name: user-1234
    securityContext:
      runAsUser: 1234
```
and it looks like real/effective don't differ:
```
busybox-65bc76cbf9-74gbv busybox real=0, effective=0
busybox-65bc76cbf9-74gbv user-1234 real=1234, effective=1234
```

^ That's certainly not an exhaustive analysis, but with containers giving so much control over the runtime environment, I wonder how much of a problem this is in practice.

**Comment by michaelschiff** [comment]
@ceason I was referring more to the first part of that quote than the particular example about UID/GID (which I agree may not be much of a problem in practice).  My comment was just that since it's possible to disable the use of `JAVA_TOOL_OPTIONS` it may be worthwhile to keep the ability to modify the command-line entrypoint directly.  However, if we are confident that this feature will be available for all use cases we are targeting, I would be fine reverting this change as well.

**Pull request merged**